contracts/Convex.sol

```diff
@@ -0,0 +1,250 @@
+    //      total supply). This value is the vp_rate variable in the meta pool's get_virtual_price() method
+    //   2. This virtual price is passed into the self._xp() method, which multiplies this by
+    //      the metacurrency token balance then divides by PRECISION. If virtual price is too
+    //      small relative to the PRECISION, the integer division is floored, returning zero.
+    //   3. This xp value of zero is passed into self._get_D(), and is used in division. We of
+    //      course cannot divide by zero, so the call reverts
+    //
+    // Given this potential failure mode, we check trigger conditions as follows:
+    //   1. First we do the balance checks since that check cannot revert
+    //   2. Next we check the virtual price of that base pool. This can still revert if the balance of
+    //      a token is too low, resulting in a zero value for xp leading to division by zero, but
+    //      because we already checked that balances are not too low this should be safe.
+    //      NOTE: There is a potential edge case where a token balance decrease is less than our 50%
+    //      threshold so the balance trigger condition is not toggled, BUT the balance is low enough
+    //      that xp is still floored to zero during integer division, resulting in a revert. More
+    //      analysis (or perhaps just a thorough understanding of Curve internals?) is needed to
+    //      ensure this is not possible.
+    //      NOTE: In a properly functioning curve market, get_virtual_price() should never revert.
+    //      Therefore, all external calls are wrapped in a try/catch, and if the call reverts then
+    //      something is wrong with the underlying protocol and we toggle the trigger
+    //   3. Lastly we check the virtual price of the meta pool for similar reasons to above
+    //
+    // For try/catch blocks, we return early if the trigger condition was met. If it wasn't, we
+    // save off the new state variable. This can result in "inconsistent" states after a trigger
+    // occurs. For example, if the first check is ok, but the second check fails, the final state
+    // of this contract will have the new state from the first check, but the prior state from the
+    // second (failed) check (i.e. not the most recent check that triggered the). This is a bit
+    // awkward, but ultimatly is not a problem
+
+    // Verify supply of Convex receipt tokens is equal to the amount of curve receipt tokens Convex
+    // can claim. Convex receipt tokens are minted 1:1 with deposited funds, so this protects
+    // against e.g. "infinite mint" type bugs, where an attacker is able to mint themselves more
+    // Convex receipt tokens than what they should receive.
+    if (IERC20(convexToken).totalSupply() != IERC20(gauge).balanceOf(staker)) return true;
+
+    // Token balance checks
+    if (checkCurveBaseBalances()) return true;
+    if (checkCurveMetaBalances()) return true;
```

Comment:
nit: could be one `if` with a `||`. Depending on how smart the compiler is, might be a minor gas optimization.

Reply:
Changed in fd73021